File: cpp/swea/swea_1247.cpp
// swea_1247 최적 경로


// v1
#include <iostream>
#include <vector>

using namespace std;

int N;
int res;
int home_x, home_y;


void dfs(int cnt, int x, int y, int distance, vector<int>& visited, vector<pair<int, int>>& position) {
	if (distance > res) return;
	if (cnt == N) {
		int tmp = distance + abs(x - home_x) + abs(y - home_y);
		if (tmp < res) {
			res = tmp;
		}
		return;
	}

	for (int i = 0; i < N; ++i) {
		if (visited[i] == 0) {
			int nx = position[i].first;
			int ny = position[i].second;
			int tmp = abs(x - nx) + abs(y - ny);
			visited[i] = 1;
			dfs(cnt + 1, nx, ny, distance + tmp, visited, position);
			visited[i] = 0;
		}
	}
}




int main() {
	cin.tie(0);
	ios_base::sync_with_stdio(0);

	int T;
	cin >> T;
	for (int t = 1; t < T + 1; ++t) {
		cin >> N;
		vector<pair<int, int>> position;
		vector<int> visited(N);
		int company_x, company_y;
		cin >> company_x >> company_y >> home_x >> home_y;

		int x, y;
		for (int n = 0; n < N; ++n) {
			cin >> x >> y;
			position.push_back({ x,y });
		}
		res = 987654321;
		dfs(0, company_x, company_y, 0, visited, position);

		cout << '#' << t << ' ' << res << '\n';
	}
}

File: python/swea/swea_1247.py
# swea_1247 최적 경로

# v1

def dfs(cnt, x, y, distance):
    global res
    if distance > res:
        return
    if cnt == N:
        dist_tmp = distance + abs(x-home[0]) + abs(y-home[1])
        if dist_tmp < res:
            res = dist_tmp
        return

    for i in range(N):
        if visited[i] == 0:
            dist_tmp = abs(x-position[i][0]) + abs(y-position[i][1])
            visited[i] = 1
            dfs(cnt+1, position[i][0], position[i][1], distance+dist_tmp)
            visited[i] = 0


T = int(input())

for t in range(1, T+1):
    N = int(input())
    position_tmp = list(map(int, input().split()))
    company = [position_tmp[0], position_tmp[1]]
    home = [position_tmp[2], position_tmp[3]]
    res = 987654321
    position = []
    for i in range(4, len(position_tmp)-1, 2):
        position.append([position_tmp[i], position_tmp[i+1]])

    visited = [0]*N
    dfs(0, company[0], company[1], 0)
    print("#{} {}".format(t, res))